### PR TITLE
Remove invalid warning during cleanup phase

### LINKED ISSUE #35061: terraform test: produces invalid warning during cleanup phase
### Terraform Version

```shell
1.7.5 / 1.8.1
```

### Terraform Configuration Files

```terraform
# main.tf

variable "input" {}

resource "tfcoremock_simple_resource" "resource" {
    string = var.input
}
```

```hcl
# main.tftest.hcl

run "test" {
    variables {
        input = "Hello, world!"
        validation = "Hello, world!"
    }

    assert {
        condition = tfcoremock_simple_resource.resource.string == var.validation
        error_message = "bad!"
    }
}
```

### Debug Output

N/A

### Expected Behavior

```bash
> terraform test
main.tftest.hcl... in progress
  run "test"... pass
main.tftest.hcl... tearing down
main.tftest.hcl... pass

Success! 1 passed, 0 failed.
```

### Actual Behavior

```bash
> terraform test
main.tftest.hcl... in progress
  run "test"... pass
main.tftest.hcl... tearing down
╷
│ Warning: Value for undeclared variable
│ 
│   on main.tftest.hcl line 5, in run "test":
│    5:         validation = "Hello, world!"
│ 
│ The module under test does not declare a variable named "validation", but it is declared in run block "test".
╵
main.tftest.hcl... pass

Success! 1 passed, 0 failed.
```

### Steps to Reproduce

1. `terraform init`
2. `terraform test`

### Additional Context

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by MicahKimel**
Fixes #35061

## Target Release

1.8.x

### NEW FEATURES | UPGRADE NOTES | ENHANCEMENTS | BUG FIXES | EXPERIMENTS

 

-  This update takes out display warnings during cleanup phase and by passing a new parameter called `action` into the GetVariables() function. This seemed to be the easiest way I could think of to know that the `GetVariables()` is being called by `Destroy()`. Let me know any feedback!

### CODE REVIEW

File: internal/backend/local/test.go [hunk 1]
```diff
@@ -1004,7 +1004,7 @@ func (runner *TestFileRunner) cleanup(file *moduletest.File) {
 // more variables than are required by the config. FilterVariablesToConfig
 // should be called before trying to use these variables within a Terraform
 // plan, apply, or destroy operation.
-func (runner *TestFileRunner) GetVariables(config *configs.Config, run *moduletest.Run, references []*addrs.Reference) (terraform.InputValues, tfdiags.Diagnostics) {
+func (runner *TestFileRunner) GetVariables(config *configs.Config, run *moduletest.Run, references []*addrs.Reference, action string) (terraform.InputValues, tfdiags.Diagnostics) {
```
Comment (liamcervante) [inline internal/backend/local/test.go:1007]:
```suggestion
func (runner *TestFileRunner) GetVariables(config *configs.Config, run *moduletest.Run, references []*addrs.Reference, includeWarnings bool) (terraform.InputValues, tfdiags.Diagnostics) {
```

I think it's better in this case to be clear about what the variable is actually controlling. It's just doing one thing at the moment and we can make the behaviour clear in the function signature with this.

**Review by liamcervante — COMMENTED** [review]
Hi @MicahKimel, thanks for this.

The overall approach looks good to me - I'd just replace the new function argument with a variable instead of the string so we can more easily document what behaviour is being changed with the argument.

Also, would you be able to add a small test for this? The testing framework is mostly testing in `command/test_test.go`. Here's an example of a test that runs `terraform init` and then `terraform test` and validates the output from the test command matches what we expect: https://github.com/hashicorp/terraform/blob/main/internal/command/test_test.go#L1955

You can add new test data in [command/testdata/test](https://github.com/hashicorp/terraform/tree/main/internal/command/testdata/test), essentially we'd just need a very simple test case that references a variable that doesn't exist. Then validate the output produces the warning during the initial run and not during the cleanup phase.

Hopefully that makes sense! Let me know if you don't have time or it doesn't make sense. We don't need a big complicated test for this, so hopefully shouldn't be too difficult. Thanks again.

File: internal/backend/local/test.go (hunk 1 above)
Comment (MicahKimel) [reply internal/backend/local/test.go]: Thank you @liamcervante! Yes `includeWarnings` is much better than my string!

**Comment by MicahKimel** [comment]
Thanks @liamcervante!! Updated solution and added test case!

File: internal/command/test_test.go [hunk 2]
```diff
@@ -1270,6 +1270,75 @@ Success! 2 passed, 0 failed.
...
+	if code != 0 {
+		t.Errorf("expected status code 0 but got %d", code)
+	}
+
+	expected := `main.tftest.hcl... in progress
+  run "test"... pass
+main.tftest.hcl... tearing down
+main.tftest.hcl... pass
+
+Success! 1 passed, 0 failed.
```
Comment (liamcervante) [inline internal/command/test_test.go:1328]: We are expecting to see a warning here during the main test phase (ie. after `run "test"... pass`), to validate that we are actually triggering the case when the warning is produced at all.

File: internal/command/testdata/test/invalid-cleanup-warnings/main.tftest.hcl [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# main.tftest.hcl
+
+run "test" {
+  variables {
+    input = "Hello, world!"
+    validation = "Hello, world!"
+  }
+  assert {
+    condition = test_resource.resource.value == var.validation
```
Comment (liamcervante) [inline internal/command/testdata/test/invalid-cleanup-warnings/main.tftest.hcl:9]:
```suggestion
    condition = test_resource.resource.value == "Hello, world!"
```

We need to make sure we are not referencing the `validation` variable anywhere - this is to make sure we do actually produce the warning during the testing phase.

**Review by liamcervante — COMMENTED** [review]
Thanks, looks good! Just a couple of very small tweaks to the test.

**Comment by MicahKimel** [comment]
Thanks! How does this look

**Review by liamcervante — COMMENTED** [review]
Your changes look perfect, but the test you've written has actually exposed another bug. Specifically around the line `(and one more similar warning elsewhere)`. We should only be adding a single diagnostic here, and not two. 

This is happening because of these lines: https://github.com/hashicorp/terraform/blob/main/internal/backend/local/test.go#L547-L551. You can see we are just needlessly re-adding and rechecking the original `variableDiags` into our main diagnostics which is causing the duplicates.

This isn't something caused by any changes in your PR but is another existing bug. Could you just delete those lines and update the new test you've added so it's not reporting it has found similar errors? This request is out of scope of the original bug, but would be nice to not add a test that is revealing another bug when it has such a simple fix.

If you don't have time we can file another issue tracking this and someone else can pick it up later - as I said it's out of scope for what you originally agreed to fix so we can merge this fix as is. Let me know what you want to do and I can approve or wait for the update as you want!

**Comment by MicahKimel** [comment]
Thanks! Removed those lines and updated the test case. Thank you for me so much with this! This is my first time contributing to an open source project and I've really appreciated your help and feedback! If this looks good, I'll have to go find a new issue to work on! :)

**Review by liamcervante — APPROVED** [review]
This looks great, thanks for your contribution!

**Comment by liamcervante** [comment]
This has been merged and backported. It will be released as part of v1.8.5.

**Pull request merged**